### ISSUE TITLE
REPL fails to start with `scalameta` as `dep`

### ISSUE THREAD

**Issue by pawelprazak** [VirtusLab/scala-cli#2582]
**Version(s)**
```
scala-cli version
Scala CLI version: 1.0.6
Scala version (default): 3.3.1
```

**Describe the bug**
`scala` REPL throws or `SIGSEGV`s when using `scalameta` as `dep`

**To Reproduce**

With default JVM:
```
scala-cli --dep org.scalameta:scalameta_2.13:4.8.14
Welcome to Scala 3.3.1 (21.0.1, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
Exception in thread "main" com.sun.jna.LastErrorException: [14] Bad address
	at com.sun.jna.Native.invokeVoid(Native Method)
	at com.sun.jna.Function.invoke(Function.java:415)
	at com.sun.jna.Function.invoke(Function.java:361)
	at com.sun.jna.Library$Handler.invoke(Library.java:270)
	at jdk.proxy1/jdk.proxy1.$Proxy0.ioctl(Unknown Source)
	at org.jline.terminal.impl.jna.osx.OsXNativePty.getSize(OsXNativePty.java:82)
	at org.jline.terminal.impl.AbstractPosixTerminal.getSize(AbstractPosixTerminal.java:60)
	at org.jline.terminal.Terminal.getBufferSize(Terminal.java:217)
	at org.jline.reader.impl.LineReaderImpl.doDisplay(LineReaderImpl.java:755)
	at org.jline.reader.impl.LineReaderImpl.<init>(LineReaderImpl.java:308)
	at org.jline.reader.LineReaderBuilder.build(LineReaderBuilder.java:121)
	at dotty.tools.repl.JLineTerminal.readLine(JLineTerminal.scala:72)
	at dotty.tools.repl.ReplDriver.readLine$1(ReplDriver.scala:157)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:167)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:172)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:211)
	at dotty.tools.repl.ReplDriver.runBody$$anonfun$1(ReplDriver.scala:185)
	at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
	at dotty.tools.repl.ReplDriver.runBody(ReplDriver.scala:185)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:172)
	at dotty.tools.repl.ReplDriver.tryRunning(ReplDriver.scala:134)
	at dotty.tools.repl.Main$.main(Main.scala:7)
	at dotty.tools.repl.Main.main(Main.scala)
```

With JVM 17:
```
scala-cli --jvm 17 --dep org.scalameta:scalameta_2.13:4.8.14

Welcome to Scala 3.3.1 (17.0.9, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000102d7ea20, pid=79350, tid=10243
#
# JRE version: OpenJDK Runtime Environment Zulu17.46+19-CA (17.0.9+8) (build 17.0.9+8-LTS)
# Java VM: OpenJDK 64-Bit Server VM Zulu17.46+19-CA (17.0.9+8-LTS, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [jna7748573358245533514.tmp+0xea20]  ffi_call+0x530
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/pprazak/repos/besom/hs_err_pid79350.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azul.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Abort trap: 6
```

logs fragment:
```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.sun.jna.Native.invokeVoid(Lcom/sun/jna/Function;JI[Ljava/lang/Object;)V+0
j  com.sun.jna.Function.invoke([Ljava/lang/Object;Ljava/lang/Class;ZI)Ljava/lang/Object;+42
j  com.sun.jna.Function.invoke(Ljava/lang/reflect/Method;[Ljava/lang/Class;Ljava/lang/Class;[Ljava/lang/Object;Ljava/util/Map;)Ljava/lang/Object;+271
j  com.sun.jna.Library$Handler.invoke(Ljava/lang/Object;Ljava/lang/reflect/Method;[Ljava/lang/Object;)Ljava/lang/Object;+390
j  jdk.proxy1.$Proxy0.ioctl(ILcom/sun/jna/NativeLong;Lorg/jline/terminal/impl/jna/osx/CLibrary$winsize;)V+27 jdk.proxy1
j  org.jline.terminal.impl.jna.osx.OsXNativePty.getSize()Lorg/jline/terminal/Size;+26
j  org.jline.terminal.impl.AbstractPosixTerminal.getSize()Lorg/jline/terminal/Size;+4
j  org.jline.terminal.Terminal.getBufferSize()Lorg/jline/terminal/Size;+1
j  org.jline.reader.impl.LineReaderImpl.doDisplay()V+8
j  org.jline.reader.impl.LineReaderImpl.<init>(Lorg/jline/terminal/Terminal;Ljava/lang/String;Ljava/util/Map;)V+357
j  org.jline.reader.LineReaderBuilder.build()Lorg/jline/reader/LineReader;+39
j  dotty.tools.repl.JLineTerminal.readLine(Lorg/jline/reader/Completer;Ldotty/tools/dotc/core/Contexts$Context;)Ljava/lang/String;+154
j  dotty.tools.repl.ReplDriver.readLine$1(Ldotty/tools/repl/JLineTerminal;Ldotty/tools/repl/State;)Ldotty/tools/repl/ParseResult;+25
j  dotty.tools.repl.ReplDriver.loop$1(Ldotty/tools/repl/JLineTerminal;Ldotty/tools/repl/State;)Ldotty/tools/repl/State;+3
j  dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(Ldotty/tools/repl/State;Ldotty/tools/repl/JLineTerminal;)Ldotty/tools/repl/State;+3
j  dotty.tools.repl.ReplDriver$$Lambda$278+0x0000000801254768.apply()Ljava/lang/Object;+12
j  dotty.tools.repl.ReplDriver.withRedirectedOutput(Lscala/Function0;)Ldotty/tools/repl/State;+30
j  dotty.tools.repl.ReplDriver.runBody$$anonfun$1(Lscala/Function0;)Ldotty/tools/repl/State;+2
j  dotty.tools.repl.ReplDriver$$Lambda$285+0x0000000801256458.apply()Ljava/lang/Object;+8
j  dotty.tools.runner.ScalaClassLoader$.asContext(Ljava/lang/ClassLoader;Lscala/Function0;)Ljava/lang/Object;+13
j  dotty.tools.repl.ReplDriver.runBody(Lscala/Function0;)Ldotty/tools/repl/State;+21
j  dotty.tools.repl.ReplDriver.runUntilQuit(Ldotty/tools/repl/State;)Ldotty/tools/repl/State;+100
j  dotty.tools.repl.ReplDriver.tryRunning()V+12
j  dotty.tools.repl.Main$.main([Ljava/lang/String;)V+20
j  dotty.tools.repl.Main.main([Ljava/lang/String;)V+4
v  ~StubRoutines::call_stub
```

**Expected behaviour**
- REPL should default to JVM 17 instead of system default
- REPL should not crash or sigsegv
- REPL should return actionable error message to the user on how to fix the problem or how to report the problem
- REPL should allow me to use `scalameta`

**Comment by pawelprazak** [comment]
~~Workaround~~ An alternative, if anyone encounters the same problem:
```
scala-cli --power repl --amm
@ import $ivy.`org.scalameta:scalameta_2.13:4.8.14`, scala.meta._
```

**Comment by Gedochao** [comment]
> **Expected behaviour**
> 
> * REPL should default to JVM 17 instead of system default

Actually, it should default to the system default if the system default is Java >17 (and you seem to have Java 21, if I'm not mistaken while looking at the logs)

> * REPL should not crash or sigsegv

agreed, I can replicate this.

Also, `--amm` is not really a workaround as much as it's using a different REPL. But yeah, ok.

**Comment by pawelprazak** [comment]
> > **Expected behaviour**
> > 
> > * REPL should default to JVM 17 instead of system default
> 
> Actually, it should default to the system default if the system default is Java >17 (and you seem to have Java 21, if I'm not mistaken while looking at the logs)
> 
> > * REPL should not crash or sigsegv
> 
> agreed, I can replicate this.
> 
> Also, `--amm` is not really a workaround as much as it's using a different REPL. But yeah, ok.

Yes, it is not a proper workaround, more of an alternative form the user POV

**Comment by pawelprazak** [comment]
> > **Expected behaviour**
> > 
> > * REPL should default to JVM 17 instead of system default
> 
> Actually, it should default to the system default if the system default is Java >17 (and you seem to have Java 21, if I'm not mistaken while looking at the logs)

OK, I trust your judgment :)

**Comment by Gedochao** [comment]
Here's a quick workaround:
```bash
export JAVA_HOME=$(cs java-home --jvm temurin:17)
echo $JAVA_HOME
# ~/Library/Caches/Coursier/arc/https/github.com/adoptium/temurin17-binaries/releases/download/jdk-17%252B35/OpenJDK17-jdk_x64_mac_hotspot_17_35.tar.gz/jdk-17+35/Contents/Home                                                                             
scala-cli --dep org.scalameta:scalameta_2.13:4.8.14
# Welcome to Scala 3.3.1 (17, Java OpenJDK 64-Bit Server VM).
# Type in expressions for evaluation. Or try :help.
#                                                                                 
# scala> 
```
If your `JAVA_HOME` points to `temurin:17` and you don't explicitly pass the `--jvm` parameter (not even to `system|17`), the REPL seems to work correctly with the `scalameta` dependency.
Still trying to figure out why that is, posting just in case that's helpful to anyone looking at the ticket.

**Comment by Gedochao** [comment]
Okay, so I'm still working on this, but I do have a better workaround, so I thought it might be worth sharing in the meantime.
The culprit is [JNA 5.13.0](https://mvnrepository.com/artifact/net.java.dev.jna/jna/5.13.0), which is a dependency of `org.scalameta:scalameta_2.13:4.8.14`.

This particular version of JNA seems to have some issues with the `arm64` architecture, which causes the weird behaviour on the REPL. Scala 3 REPL itself is using JNA of a lower version: `5.3.1`, which seems to play nice with `arm64`. However, as `scalameta` drags a newer version to the REPL classpath, it all goes to hel.

So from the perspective of Scala CLI, this is the minimised bug:

```bash
scala-cli repl --dep net.java.dev.jna:jna:5.13.0        
# Welcome to Scala 3.3.1 (17, Java OpenJDK 64-Bit Server VM).
# Type in expressions for evaluation. Or try :help.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000105d36a20, pid=65565, tid=6147
#
# JRE version: OpenJDK Runtime Environment Temurin-17+35 (17.0+35) (build 17+35)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17+35 (17+35, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [jna1869522016086249606.tmp+0xea20]  ffi_call+0x530
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/pchabelski/Downloads/hs_err_pid65565.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
# [1]    65565 abort      scala-cli repl --dep net.java.dev.jna:jna:5.13.0
```
Do mind that when `JNA 5.3.1` is passed, all is fine.
```bash
scala-cli repl --dep net.java.dev.jna:jna:5.3.1 
# Welcome to Scala 3.3.1 (17, Java OpenJDK 64-Bit Server VM).
# Type in expressions for evaluation. Or try :help.
#                                                                                                                                                                          
#scala> 
```
It is also safe to use `5.13.0` (and `scalameta` as an extension) on the Scala 2.13 REPL (and `scalameta` from the original reproduction is built for 2.13)
```bash
scala-cli repl -S 2.13 --dep net.java.dev.jna:jna:5.13.0
# Welcome to Scala 2.13.12 (OpenJDK 64-Bit Server VM, Java 17).
# Type in expressions for evaluation. Or try :help.
# 
# scala> 
```

Another workaround is to explicitly use non-arm64 Java, even if you are on an arm64 system (which I'm assuming would be an Apple M1 machine in most cases at this time). 
The issue is gone when using Java build for AMD64:
```bash
scala-cli repl --dep net.java.dev.jna:jna:5.13.0 --jvm zulu:17 --jvm-index-arch amd64
Welcome to Scala 3.3.1 (17.0.9, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                                                         
scala> 
```
Switching to amd64 Java when working with `scalameta` in a REPL on an M1 seems like the sanest workaround for the time being.

Now, as for why the previous workaround seems to be working: Coursier CLI seems to return amd64 Java when I call `cs java-home --jvm temurin:17` from the command line (even though I am myself working on arm64). 
This confused me to no end earlier, as I have been unknowingly testing this on an amd64 Java because of it.

Finally, one can simply exclude the troublesome JNA 5.13.0 (and be left with the version the Dotty compiler uses, which is safe for aarch64)
```bash
scala-cli --dep 'org.scalameta:scalameta_2.13:4.8.14,exclude=net.java.dev.jna%jna'
# Welcome to Scala 3.3.1 (17, Java OpenJDK 64-Bit Server VM).
# Type in expressions for evaluation. Or try :help.
#                                                                                                                                                                           
# scala> 
```

TL;DR Either exclude JNA from the `scalameta` dependency with the `exclude` classifier or use `amd64` Java as a workaround on `arm64` machines, you can force the Java architecture with the `--jvm-index-arch amd64` option. Alternatively, use the Scala 2.13 REPL with `-S 2.13` (or Ammonite REPL with `--amm`).

**Comment by pawelprazak** [comment]
wow, kudos for tracking this dependency hell down, and thank you for the workarounds!
